src/xrpld/app/tx/detail/InvariantCheck.cpp

```diff
@@ -2170,4 +2179,911 @@ ValidAMM::finalize(
     return true;
 }
 
+//------------------------------------------------------------------------------
+
+ValidVault::Vault
+ValidVault::Vault::make(SLE const& from)
+{
+    XRPL_ASSERT(
+        from.getType() == ltVAULT,
+        "ValidVault::Vault::make : from Vault object");
+
+    ValidVault::Vault self;
+    self.key = from.key();
+    self.asset = from.at(sfAsset);
+    self.pseudoId = from.getAccountID(sfAccount);
+    self.shareMPTID = from.getFieldH192(sfShareMPTID);
+    self.assetsTotal = from.at(sfAssetsTotal);
+    self.assetsAvailable = from.at(sfAssetsAvailable);
+    self.assetsMaximum = from.at(sfAssetsMaximum);
+    self.lossUnrealized = from.at(sfLossUnrealized);
+    return self;
+}
+
+ValidVault::Shares
+ValidVault::Shares::make(SLE const& from)
+{
+    XRPL_ASSERT(
+        from.getType() == ltMPTOKEN_ISSUANCE,
+        "ValidVault::Shares::make : from MPTokenIssuance object");
+
+    ValidVault::Shares self;
+    self.share = MPTIssue(
+        makeMptID(from.getFieldU32(sfSequence), from.getAccountID(sfIssuer)));
+    self.sharesTotal = from.at(sfOutstandingAmount);
+    self.sharesMaximum = from[~sfMaximumAmount].value_or(maxMPTokenAmount);
+    return self;
+}
+
+void
+ValidVault::visitEntry(
+    bool isDelete,
+    std::shared_ptr<SLE const> const& before,
+    std::shared_ptr<SLE const> const& after)
+{
+    // If `before` is empty, this means an object is being created
+    // If `after` is empty, this means an object is being deleted
+    // Otherwise it's a modification of an object.
+    XRPL_ASSERT(
+        before != nullptr || after != nullptr,
+        "ripple::ValidVault::visitEntry : some object is available");
+
+    // `Number balance` will capture the difference (delta) between "before"
+    // state (zero if created) and "after" state (zero if destroyed), so the
+    // invariants can validate that the change in account balances matches the
+    // change in vault balances, stored to deltas_ at the end of this function.
+    Number balance{};
+
+    // By default do not add anything to deltas
+    std::int8_t sign = 0;
+    if (before)
+    {
+        switch (before->getType())
+        {
+            case ltVAULT:
+                beforeVault_ = Vault::make(*before);
+                break;
```

Comment:
This whole class has a "fatal flaw". (Well, not really fatal, but very significant.)

You're assuming that only one `Vault` will be affected. What if there are more? You're going to, for example, overwrite `befaultVault_` here.

You've got the same problem when you're looking up assets and shares in `finalize`, where you return the first matching item found. What if there are multiple matches?

It should be impossible, but _*that*_ is what invariants are here to confirm.

All of these structures and functions should be using collections (vector, set, whatever) instead of single values. You could have a simple check after each one that's like
```
if (beforeVault_.size() > 1)
{
    XRPL_ASSERT(enforce, ...);
    return !enforce;
}
```

Reply:
Thanks, that's great feedback. Making it work is actually simple (since `optional` is basically a collection of at-most-one)

Reply:
> Thanks, that's great feedback. Making it work is actually simple (since `optional` is basically a collection of at-most-one)

I figured it would be. 😄